crypto/s2n_pkey.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <openssl/evp.h>
+#include <openssl/x509.h>
+
+#include "error/s2n_errno.h"
+
+#include "crypto/s2n_pkey.h"
+
+#include "utils/s2n_safety.h"
+
+int s2n_pkey_zero(struct s2n_pkey *pkey) 
+{
+    pkey->sign = NULL;
+    pkey->verify = NULL;
+    pkey->encrypt = NULL;
+    pkey->decrypt = NULL;
+    pkey->match = NULL;
+    pkey->free = NULL;
+    return 0;
+}
+
+int s2n_pkey_init_for_type(struct s2n_pkey *pkey, s2n_cert_type cert_type)
+{
+    switch(cert_type){
+    case S2N_CERT_TYPE_RSA_SIGN:
+        GUARD(s2n_rsa_pkey_init(pkey));
+        break;
+    case S2N_CERT_TYPE_ECDSA_SIGN:
+        GUARD(s2n_ecdsa_pkey_init(pkey));
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
+    }
+    return 0;
+}
+
+int s2n_pkey_sign(const struct s2n_pkey *pkey, struct s2n_hash_state *digest, struct s2n_blob *signature)
+{
+    notnull_check(pkey->sign);
+    
+    return pkey->sign(pkey, digest, signature);
+}
+
+int s2n_pkey_verify(const struct s2n_pkey *pkey, struct s2n_hash_state *digest, struct s2n_blob *signature)
+{
+    notnull_check(pkey->verify);
+    
+    return pkey->verify(pkey, digest, signature);
+}
+
+int s2n_pkey_encrypt(const struct s2n_pkey *pkey, struct s2n_blob *in, struct s2n_blob *out)
+{
+    notnull_check(pkey->encrypt);
+
+    return pkey->encrypt(pkey, in, out);
+}
+
+int s2n_pkey_decrypt(const struct s2n_pkey *pkey, struct s2n_blob *in, struct s2n_blob *out)
+{
+    notnull_check(pkey->decrypt);
+
+    return pkey->decrypt(pkey, in, out);
+}
+
+int s2n_pkey_match(const struct s2n_pkey *pub_key, const struct s2n_pkey *priv_key)
+{
+    notnull_check(pub_key->match);
+
+    return pub_key->match(pub_key, priv_key);
+}
+
+int s2n_pkey_free(struct s2n_pkey *pkey)
+{
+    if (pkey == NULL || pkey->free == NULL) {
+        return 0;
```

Comment:
Hmm.. I see your point. `s2n_pkey_zero` is actually called just during initialization of the key. For example, in `s2n_config_add_cert_chain_and_key`, we initialize several pointer members to NULL, and `s2n_pkey_zero` was added to achieve that for the key. It's also called when parsing the client certificate. I don't think I can move pkey_zero into pkey_free, but I think I could change the name to more strongly imply it is used to initialize rather than destroy.